design-proposals/klip-7.md

```diff
@@ -0,0 +1,176 @@
+
+#### Execution Plans
+To address this, I propose we build our own representation of the topology (called an Execution Plan) that provides details about what the streams operators are actually doing. Then, in our compatibility verification tests, we can verify that Execution Plans are kept consistent across versions.
+
+The Execution Plan is a DEG of Execution Steps. Each Execution Step represents a call to Kafka Streams, along with context about what the resulting Streams operator is doing. We can build the Execution Plan at the SchemaKStream layer.
+
+For example, the call to SchemaKStream.select would look something like:
+
+```
+public SchemaKStream<K> select(
+    final List<SelectExpression> selectExpressions,
+    final QueryContext.Stacker contextStacker,
+    final ProcessingLogContext processingLogContext) {
+  final ExecutionStep step = new StreamSelect(this.executionStep, selectExpressions);
+  return new SchemaKStream<>(
+      step.build(),
+      ksqlConfig,
+      functionRegistry,
+      contextStacker.getQueryContext(),
+      step
+  );
+}
+```
+
+Where StreamSelect looks like:
+
+```
+public class StreamSelect implements ExecutionStep {
+  ...
+  private final ExecutionStep source;
+  private final List<SelectExpression> selectExpressions;
+
+  @JsonCreator
+  public StreamSelect(
+      ...
+      @JsonProperty("source") final ExecutionStep source,
+      @JsonProperty("selectExpressions") final List<SelectExpression> selectExpressions) {
+    this.properties = Objects.requireNonNull(properties, "properties");
+    this.source = Objects.requireNonNull(source, "source");
+    this.selectExpressions = ImmutableList.copyOf(selectExpressions);
+  }
+
+  public List<SelectExpression> getSelectExpressions() {
+    return selectExpressions;
+  }
+
+  @Override
+  public List<ExecutionStep> getSources() {
+    return Collections.singletonList(source);
+  }
+
+  @Override
+  public KStream build(final KStream input) {
+      final Selection selection = new Selection(selectExpressions);
+      return kstream.mapValues(selection.getSelectValueMapper()),
+  }
+}
+```
+
+I've written up a basic implementation of Execution Plans which includes all the Execution Steps we'd need, and can be found here (https://github.com/confluentinc/ksql/blob/ksql-execution-plans/ksql-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java) (steps are under ksql-engine/src/main/java/io/confluent/ksql/structured/execution):
+
+#### Serializability
+The Execution Plan needs to be JSON-serializable. This way, we can persist the plan generated for each case in our Query Translation tests, and verify that it is kept the same across versions.
```

Comment:
JSON kinda sucks as a serialization format (but is great for readability!) and these plans might end up being rather large - I'm not too worried in the short term but this is something we might want to consider.

Reply:
That's fair. I didn't put much thought into the choice of JSON here beyond it being the easiest thing to implement. Will think about this some more.

Reply:
The data persisted to the command topic is likely compressed anyway. (Depending on Kafka broker / producer properties)

We could always use BSON.